### PR TITLE
Return Pydantic models from API Endpoints

### PR DESCRIPTION
**Pull request by jmsmkn**
This PR adds a quality of life improvement by being explicit about the data returned from Grand Challenge. It allows users to directly use attributes on the returned instances rather than looking though dictionaries, and also validates the type of data returned.

We may get a lot of errors as we find the limitations of how our API is defined, but it works for our integration tests right now.

At the moment the inferred types are not correct, but they should follow later.

See https://github.com/DIAGNijmegen/rse-roadmap/issues/213

### CODE REVIEW

**Comment by chrisvanrun** [comment]
Cool!

File: gcapi/models.py [hunk 1]
```diff
@@ -0,0 +1,2129 @@
+# generated by datamodel-codegen:
```
Comment (jmsmkn) [inline gcapi/models.py:1]: Auto-generated by `generate_models.py`, doesn't need to be reviewed thoroughly, maybe just the overall structure of using pydantic dataclasses.

File: gcapi/sync_async_hybrid_support.py [hunk 2]
```diff
@@ -237,8 +237,10 @@ def __call__(self, *args, **kwargs):
 
 GENERATOR_MARKER = "__gcapi_generator"
 
+RT = TypeVar("RT")
```
Comment (jmsmkn) [inline gcapi/sync_async_hybrid_support.py:240]: Return Type

**Comment by jmsmkn** [comment]
Eventually we can add it to a GitHub actions workflow that creates a PR when we want to update the models. The resulting file should be reviewed so shouldn't be an automated build command. Could go into a Makefile, but the `gcapi/models.py` file will already exist so it would be PHONY anyway. For those reasons I thought a loose script was fine for now.

**Comment by pkcakeout** [comment]
One more question then - should I give this a full review?

> We may get a lot of errors as we find the limitations of how our API is defined, but it works for our integration tests right now.

... or does that mean you would like to add the generator for the correct `XXXClient`-classes as well first?

**Review by pkcakeout — COMMENTED** [review]
I think there are the minor improvements (see code suggestions) could be applied and makes this PR better.

And then the `ast` thing that likely is overkill. But I did it, so if you want it, we can use it. Likely is more robust!?

File: gcapi/apibase.py [hunk 3]
```diff
@@ -93,7 +158,7 @@ def iterate_all(self, params=None):
             yield from current_list
             offset += req_count
 
-    def detail(self, pk=None, **params):
+    def detail(self, pk=None, **params) -> Generator[T, Dict[Any, Any], T]:
```
[the remark this answers is not in the document]
Comment (jmsmkn) [reply gcapi/apibase.py:161]: I am glad it's not just me!

File: generate_models.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            target_python_version=PythonVersion.PY_38,
+        )
+
+        with open(Path(__file__).parent / "gcapi" / "models.py", "w") as f:
+            f.write(
+                output.read_text().replace(
+                    "from dataclasses import dataclass, field",
+                    "from pydantic.dataclasses import dataclass",
+                )
+            )
```
[the remark this answers is not in the document]
Comment (jmsmkn) [reply generate_models.py]: I expect it all to be fragile, hence pinning to patch versions of the model generator. The tests will fail if the replacement doesn't work.

File: gcapi/apibase.py [hunk 5]
```diff
@@ -38,17 +52,61 @@ def __call__(
...
+        limit: int,
+        total_count: int,
+        results: List[T],
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self._offset = offset
+        self._limit = limit
+        self._total_count = total_count
+        self._results = results
```
[the remark this answers is not in the document]
Comment (pkcakeout) [reply gcapi/apibase.py:69]: I do not think this is super-important, but probably a bit cleaner. :shrug: As I said - no dealbreaker.

**Comment by jmsmkn** [comment]
> I think this can be merged... once the precommit issue is fixed! :-) Pretty neat feature!

Thanks for the reminder, should be fine now.

**Review by pkcakeout — APPROVED** [review]
:seal: of :+1:

**Pull request merged**